test/java/org/chromium/net/testing/TestUploadDataProvider.java

```diff
@@ -0,0 +1,257 @@
+package org.chromium.net.testing;
+
+import android.os.ConditionVariable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.ArrayList;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.chromium.net.UploadDataProvider;
+import org.chromium.net.UploadDataSink;
+
+/**
+ * An UploadDataProvider implementation used in tests.
+ */
+public class TestUploadDataProvider extends UploadDataProvider {
```

Comment:
`UploadDataProviderTest`?

Reply:
In Cronet, only CronetTestRule has its test. TBH, I suspect this will stay like that...